app/models/labware_creators/multi_stamp_tubes_using_tube_rack_scan.rb

```diff
@@ -148,6 +149,25 @@ def tubes_must_have_active_requests_of_expected_type
       end
     end
 
+    # Validates that the tubes in the parent_tubes hash contains at the source tube.
+    # If the provided file does not contain the source tube, an error message is added to the errors object.
+    def tubes_must_contain_source_tube
+      return unless file_valid?
+
+      # parent tube should be LRC Bank Seq or LRC Bank Spare
+      parent_tube_barcode = labware.barcode.prefix + labware.barcode.number.to_s
```

Comment:
What do you get from just `labware.barcode`, is that not usable?

Reply:
OK so I think the most important question is:

1. Does this approach work when the tube barcodes are of the expected type for this specific pipeline? In this case, they will be FluidX barcode, which is a foreign barcode type in SS. If you run the scRNA Core Cell Extraction test in the Integration Suite it should create you some LRC Bank Seq and Spare tubes that have realistic barcodes (of the format LLNNNNNNNN where L is letter and N is number). Then you can test your new feature using more realistic barcodes. Would be interested to see what the 'ean13', 'number' and 'prefix' attributes return for a FluidX barcode.

I was also wondering:
2. Can we make it work for any barcode type? 

However, I've just realised this isn't really relevant, because this feature is specifically parsing a tube rack scan file, so we're firmly in the realm of FluidX tube racks here - we don't need to worry about other barcode types (like NT...)

Reply:
OK great thanks for that! I would go with 'machine' then.

Reply:
Discussed briefly in stand up, but I'll put my understanding here as well:

I think we used to use ean13 barcodes for everything. ean13 has to be 13 digits, which is super difficult for a human to read, so there's something (in https://github.com/sanger/sanger_barcode_format I think) that converts between the ean13 'machine' format and a more human readable format.

When we're using code39, code128 or 'foreign' barcode formats (like in this case), there's no distinction between the 'machine' and 'human readable' format - the barcode can contain letters so there's no point converting it to a more human-readable form.

We do still use ean13 (Sanger barcode assigned to tubes), so we need to keep that distinction between machine and human. So I guess there was a choice for the other barcode formats as to what to put in the 'machine' attribute for the other barcode types, and the decision was to return the barcode whether you ask for human or machine. It does kind of make sense, because it's what is read by the scanner.

There's my brain dump, hope it makes sense 😆